### DISCUSSION TITLE
Homebrew Node download errors

### DISCUSSION THREAD

**Discussion by swissflyer** [Homebrew/discussions#284] [Getting started]
Hi, I recently downloaded Homebrew and now am attempting to install Node. I ran into a few errors when running 
"brew install node"

Pouring node-15.4.0.catalina.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/node
Target /usr/local/bin/node
already exists. You may want to remove it:
  rm '/usr/local/bin/node'

Additionally, at the end it printed

Warning: The post-install step did not complete successfully
You can try again using `brew postinstall node`
==> Summary
🍺  /usr/local/Cellar/node/15.4.0: 3,273 files, 55.4MB
Error: /Library/Java/JavaVirtualMachines/liberica-jdk-8-full.jdk/Contents/Home is not a valid keg

I am new to Homebrew and would appreciate any help on the matter. Thank you!

**Comment by carlocab** [comment]
What does `brew info node` show you? My guess is you have an old version of node that didn't uninstall properly.

**Reply by swissflyer** [reply]
node: stable 15.4.0 (bottled), HEAD
Platform built on V8 to build network applications
https://nodejs.org/
/usr/local/Cellar/node/15.4.0 (3,273 files, 55.4MB)
  Poured from bottle on 2020-12-14 at 09:25:41
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/node.rb
License: MIT
==> Dependencies
Build: pkg-config ✔, python@3.9 ✘
Required: icu4c ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 368,677 (30 days), 1,182,422 (90 days), 4,412,856 (365 days)
install-on-request: 268,315 (30 days), 859,135 (90 days), 3,212,861 (365 days)
build-error: 0 (30 days)

**Reply by swissflyer** [reply]
I have no idea if this may be a factor, but when I run "echo $PATH" I get the following 
/Library/Frameworks/Python.framework/Versions/3.7/bin:/Library/Frameworks/Python.framework/Versions/3.7/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/VMware Fusion.app/Contents/Public:/usr/local/share/dotnet:~/.dotnet/tools:/Library/Frameworks/Mono.framework/Versions/Current/Commands

**Reply by carlocab** [reply]
Do you have Homebrew Python installed too? If so, I suspect that may be broken as well (hence the ❌ next to it under `brew info`). If not, then you're fine here. Do you know why your `PATH` is set that way? You may have installed Python from python.org, or something.

What does `readlink /usr/local/bin/node` say? Then do it again with `-f` and `-e` flags (separately, so that's three times).

I suspect you had a pre-existing installation of node (not by Homebrew), so that's causing you problems here.

**Reply by swissflyer** [reply]
All three instances of `readlink /usr/local/bin/node` with their respective flags came back with nothing

As for my PATH, I had to use Tensorflow awhile ago that at the time only supported an outdated python version. So, I think I may have downloaded directly from python.org

**Reply by carlocab** [reply]
Hmmm. Ok. What about `file /usr/local/bin/node`?

**Reply by swissflyer** [reply]
/usr/local/bin/node: Mach-O 64-bit executable x86_64

**Reply by carlocab** [reply]
Ok. You've got a pre-existing installation of Node via some other source. The simplest thing to do might be to just uninstall it and reinstall Homebrew Node.

In case you don't want to uninstall your currently installed Node, then all you really need is to move all files in `/usr/local` that clash with Homebrew Node. But tracking them down can get tedious.

**Reply by swissflyer** [reply]
Okay thank you. However when I run `brew uninstall node` I then get again "Error: /Library/Java/JavaVirtualMachines/liberica-jdk-8-full.jdk/Contents/Home is not a valid keg"

**Reply by carlocab** [reply]
Hmmm. Ok. I think the problem is that you have a symlink to that directory somewhere in `/usr/local/{bin,etc,include,lib,share}`. You may need to track it down and remove it. `ls -l | grep 'JavaVirtual'` or something like that may help.

Otherwise, it's possible that a symlink to that directory somehow ended up in `/usr/local/Cellar`.

**Reply by swissflyer** [reply]
I just looked through all of the respective locations and could not any find JavaVirtual Reference. Also, the only thing I found in `/usr/local/Cellar` was node, but no reference to JavaVirtual.

**Reply by carlocab** [reply]
Ugh. What does `brew config` and `brew doctor` say?

**Reply by swissflyer** [reply]
`brew config`
HOMEBREW_VERSION: 2.6.2
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1a52862e6d24cd3bf87bec557a8886a1a825ee3c
Last commit: 6 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: d914aab917710e94170526641d16414cb6a2083e
Core tap last commit: 60 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 11.0 build 1103
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 15.0.1, 1.8.0_275-b01
macOS: 10.15.7-x86_64
CLT: 1103.0.32.62
Xcode: N/A


`brew doctor`
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you an update run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/more/.


Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7-config
  /Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7m-config
  /Library/Frameworks/Python.framework/Versions/3.7/bin/python3-config

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libtsk.19.dylib
  /usr/local/lib/libtsk_jni.0.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/node/cppgc/allocation.h
  /usr/local/include/node/cppgc/common.h
....
  /usr/local/include/node/zconf.h
  /usr/local/include/node/zlib.h

Warning: Unbrewed .la files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected .la files:
  /usr/local/lib/libtsk.la
  /usr/local/lib/libtsk_jni.la

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libtsk.a
  /usr/local/lib/libtsk_jni.a

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause brews that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  node
  python@3.9

Warning: Homebrew's sbin was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting the PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
Error: /Library/Java/JavaVirtualMachines/liberica-jdk-8-full.jdk/Contents/Home is not a valid keg

**Reply by carlocab** [reply]
Yea, looks like you do have Homebrew Python installed, and that installation is broken too. Almost certainly from a pre-existing Python installation. You may wish to try the same as with node (uninstall/move pre-existing install, reinstall Python).

Unfortunately, that doesn't really tell me how to fix `node`. I'll try to think about it. Sorry I can't be more helpful.

**Reply by carlocab** [reply]
So I guess the difficulty is that Homebrew seems to think that whatever you have in `/Library/Java/JavaVirtualMachines/liberica-jdk-8-full.jdk/Contents/Home` is a keg for some reason. I didn't even think that location was writeable due to SIP.

If you're truly desperate, one thing I would personally try is to move 
```
/Library/Java/JavaVirtualMachines/liberica-jdk-8-full.jdk/Contents/Home
```
to a different location then do `brew cleanup --prune-prefix --dry-run`. The `brew cleanup` should show you any broken symlinks moving the directory above will have created. You can then do `brew cleanup --prune-prefix` to remove them, if you wish.

While I would personally try this, I also want to mention that I don't know for sure that this definitely will not break something, so tread carefully.

**Reply by carlocab** [reply]
Btw, it looks like these are the uninstallation instructions for the JDK you have installed: https://bell-sw.com/pages/liberica_install_guide-8u272/#how-to-uninstall-liberica

**Reply by swissflyer** [reply]
Thank you very much for the help! I truly appreciate it

**Reply by carlocab** [reply]
Did you get it fixed?

**Reply by swissflyer** [reply]
Not yet...I got the invalid keg error to go away. However am still working on linking /usr/local/bin etc....
I found out that the python variable in my path was from my ~/zprofile from downloading python at some time. However my path is still `/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/VMware Fusion.app/Contents/Public:/usr/local/share/dotnet:~/.dotnet/tools:/Library/Frameworks/Mono.framework/Versions/Current/Commands` and I need to find out why

**Reply by carlocab** [reply]
I see. Well, good to know you're making progress. Good luck!
